ambry-cloud/src/main/java/com/github/ambry/cloud/BackupIntegrityMonitor.java

```diff
@@ -210,6 +215,109 @@ private boolean stopLocalStore(AmbryPartition partition) {
     }
   }
 
+  /**
+   * Compares metadata received from servers with metadata received from Azure cloud
+   * @param serverReplica
+   * @param cloudReplica
+   * @param azureBlobs
+   * @param partitionBackedUpUntil
+   */
+  void compareMetadata(RemoteReplicaInfo serverReplica, RemoteReplicaInfo cloudReplica,
+      HashMap<String, MessageInfo> azureBlobs, long partitionBackedUpUntil) {
+    long partitionId = serverReplica.getReplicaId().getPartitionId().getId();
+    long scanStartTime, scanEndTime = serverReplica.getReplicatedUntilTime();
+    DateFormat formatter = new SimpleDateFormat(VcrReplicationManager.DATE_FORMAT);
+    try {
+      serverScanner.setAzureBlobInfo(azureBlobs, partitionBackedUpUntil);
+      serverScanner.addRemoteReplicaInfo(serverReplica);
+      logger.info("[BackupIntegrityMonitor] Queued peer server replica for scan [{}]", serverReplica);
+
+      while (scanEndTime < partitionBackedUpUntil) {
+        scanStartTime = serverReplica.getReplicatedUntilTime();
+        serverScanner.replicate();
+        scanEndTime = serverReplica.getReplicatedUntilTime();
+        if (scanEndTime - scanStartTime > SCAN_MILESTONE) {
+          // Print progress, if a SCAN_MILESTONE's worth of data has been received from server
+          logger.info("[BackupIntegrityMonitor] Scanned {} num_server_blobs from peer server replica {} until {}, stop at {}",
+              serverScanner.getNumBlobScanned(), serverReplica, formatter.format(scanEndTime),
+              formatter.format(partitionBackedUpUntil));
+        }
+      }
+      logger.info("[BackupIntegrityMonitor] Scanned {} num_server_blobs from peer server replica {}",
+          serverScanner.getNumBlobScanned(), serverReplica);
+
+      serverScanner.printKeysAbsentInServer(serverReplica);
+      ReplicationMetrics rmetrics = new ReplicationMetrics(cluster.getMetricRegistry(), Collections.emptyList());
+      logger.info("[BackupIntegrityMonitor] Verified cloud backup partition-{} against peer server replica {}"
+              + ", num_azure_blobs = {}, num_server_blobs = {}, num_mismatches = {}",
+          partitionId, serverReplica, ((RecoveryToken) cloudReplica.getToken()).getNumBlobs(),
+          serverScanner.getNumBlobScanned(),
+          rmetrics.backupIntegrityError.getCount());
+    } catch (Throwable e) {
+      metrics.backupCheckerRuntimeError.inc();
+      logger.error("[BackupIntegrityMonitor] Failed to verify server replica {} due to {}",
+          serverReplica, e.getMessage());
+      // Swallow all exceptions and print a trace for inspection, but do not kill the job
+      e.printStackTrace();
+    }
+
+    try {
+      if (serverReplica != null) {
+        serverScanner.removeRemoteReplicaInfo(serverReplica);
+      }
+    } catch (Throwable e) {
+      metrics.backupCheckerRuntimeError.inc();
+      logger.error("[BackupIntegrityMonitor] Failed to dequeue server replica {} due to {}",
```

Comment:
you can add exception to the end of the log error method, and in your message string, you don't have to reserve a {} for the exception. In this way, logger would be able to print out the message of the exception as well as the stack trace.